### ISSUE TITLE
Some PRNG seeds causing zero gradients

### ISSUE THREAD

**Issue by Eliacus** [google/jax#19268]
### Description

Hey,

I have a minimal example here of an MLP (using flax) training script that, for some seeds, doesn't see any convergence (loss is constant and gradients are all zero). I've tried digging deeper but can't seem to figure out why this is happening

I've included a couple of bad seeds, in which the issue occurs, and some random seeds where I don't see it.

Also, changing the second dimension of the input from 16 to 17 mitigates the issue (line 82, I left a comment there).

 
```python
from typing import Any, Callable, Dict, List, Sequence, Tuple, Union

import flax.linen as nn
import jax
import jax.numpy as jnp
import numpy as np
import optax
from flax.training.train_state import TrainState
from torch.utils.data import DataLoader, Dataset

batch_size = 64
learning_rate = 0.0001
weight_decay = 0.0001


# Bad seeds (results in gradients becoming 0)
seed = 620
# seed = 987
# seed = 218

# Good keys
# seed = 252
# seed = 846
# seed = 189

class MultiLayerPerceptron(nn.Module):
    mlp_hidden_layer_sizes: Sequence[int]
    mlp_output_layer: int

    def setup(self) -> None:
        self.hidden_layers = [nn.Dense(mlp_layer) for mlp_layer in self.mlp_hidden_layer_sizes]
        self.output_layer = nn.Dense(self.mlp_output_layer)

    def __call__(self, x: jnp.ndarray, *args: Any, **kwargs: Any) -> jnp.ndarray:
        for layer in self.hidden_layers:
            x = nn.relu(layer(x))
        x = nn.relu(self.output_layer(x))
        return x


class JaxDataset(Dataset):
    """Pytorch Dataset wrapper for jax arrays"""

    def __init__(self, x: np.ndarray, y: np.ndarray) -> None:
        self.x = x
        self.y = y

    def __len__(self) -> int:
        return len(self.x)

    def __getitem__(self, index: int) -> Tuple[np.ndarray, np.ndarray]:
        return self.x[index], self.y[index]


def numpy_collate_fn(
    batch: Union[Tuple[np.ndarray], List[np.ndarray], Tuple[Tuple], List[Tuple], Tuple[List], List[List]],
) -> np.ndarray:
    """Collate function used to make pytorch Dataloader return jax-compatible np arrays"""
    if isinstance(batch[0], np.ndarray):
        return np.stack(batch)
    if isinstance(batch[0], (tuple, list)):
        transposed = zip(*batch)
        return [numpy_collate_fn(samples) for samples in transposed]
    return np.array(batch)


def sum_of_gradients(gradients):
    sum_grads = 0
    for k, v in gradients["params"].items():
        sum_grads += np.sum(v["kernel"])
        sum_grads += np.sum(v["bias"])
    return sum_grads


def train_model():
    rng = jax.random.PRNGKey(seed)
    rng, key = jax.random.split(rng)

    inputs = jnp.ones((22999, 16)) # Changing 16->17 removes the issue
    labels = jnp.ones((22999, 1))

    train_dataset = JaxDataset(np.asarray(inputs), np.asarray(labels))
    train_dataloader = DataLoader(
        train_dataset,
        batch_size=batch_size,
        drop_last=False,
        collate_fn=numpy_collate_fn,
        num_workers=0,
    )

    network = MultiLayerPerceptron(mlp_hidden_layer_sizes=(64, 32), mlp_output_layer=1)
    params = network.init(key, inputs)
    state = TrainState.create(
        apply_fn=network.apply,
        params=params,
        tx=optax.adamw(learning_rate=learning_rate, weight_decay=weight_decay),
    )

    # @jit
    def mse(params: Dict, X: jnp.ndarray, y: jnp.ndarray) -> jnp.ndarray:
        err = state.apply_fn(params, X) - y
        return jnp.mean(jnp.square(err)) / 2

    loss_grad_fn = jax.value_and_grad(mse)

    # @jit
    def train_step(state: TrainState, batch_x: jnp.ndarray, batch_y: jnp.ndarray) -> Tuple[TrainState, float]:
        mse_loss, gradients = loss_grad_fn(state.params, batch_x, batch_y)
        print(f"batch loss: {mse_loss}")
        print(f"Sum of gradients: {sum_of_gradients(gradients)}")
        state = state.apply_gradients(grads=gradients)
        return state, mse_loss

    for batch_x, batch_y in train_dataloader:
        state, mse_loss = train_step(state, batch_x, batch_y)


if __name__ == "__main__":
    train_model()
```

### What jax/jaxlib version are you using?

jax==0.4.23, jaxlib==0.4.23+cuda11.cudnn86, jaxopt==0.8.1, flax==0.7.5

### Which accelerator(s) are you using?

GPU

### Additional system info?

Python 3.11.6, Ubuntu 20.04.6 LTS

### NVIDIA GPU info

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.147.05   Driver Version: 525.147.05   CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            On   | 00000001:00:00.0 Off |                  Off |
| N/A   45C    P0    27W /  70W |  12477MiB / 16384MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

```

**Comment by leonl42** [comment]
I think the problem is that you apply relu to the output of the last layer. For the given seeds and input sizes the output is just < 0 and thus there is no gradient.

**Comment by Eliacus** [comment]
Yep that's correct, thanks! It seems this was causing gradients to always be zero and no learning to occur. Removing the last relu or using leaky relu mitigated the issue
